File: backend/app/api/routes/github.py
"""
GitHub PR and status routes.
POST /api/v1/github/pr          — create branch, commit, open PR
GET  /api/v1/github/status/{run_id} — poll workflow run status
"""

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel
from typing import Optional

from app.api.deps import get_required_token
from app.core.exceptions import AppException
from app.github_client.repo_client import create_pr
from app.github_client.poll_client import get_run_status

router = APIRouter(prefix="/github")


class CreatePRRequest(BaseModel):
    """Request body for PR creation."""
    repo_url: str
    yaml: str
    branch_name: str = "ci/add-pipeline"


@router.post("/pr")
async def create_pull_request(
    request: CreatePRRequest,
    token: str = Depends(get_required_token),
):
    """
    Create a new branch, commit the generated workflow YAML,
    and open a Pull Request on the user's repository.
    """
    try:
        result = create_pr(
            token=token,
            repo_url=request.repo_url,
            yaml_content=request.yaml,
            branch_name=request.branch_name,
        )
        return {
            "success": True,
            "message": "Pull request created successfully.",
            "data": result,
        }
    except ValueError as e:
        raise AppException(message=str(e), status_code=400)
    except Exception as e:
        raise AppException(
            message=f"Failed to create pull request: {str(e)}",
            status_code=500,
        )


@router.get("/status/{run_id}")
async def get_workflow_status(
    run_id: str,
    repo_url: str = Query(..., description="GitHub repository URL"),
    token: str = Depends(get_required_token),
):
    """
    Poll the status of a GitHub Actions workflow run.
    Returns partial results if the run is still in progress.
    """
    try:
        status = get_run_status(
            token=token,
            repo_url=repo_url,
            run_id=int(run_id),
        )
        return {
            "success": True,
            "message": "Status retrieved.",
            "data": status,
        }
    except ValueError as e:
        raise AppException(message=str(e), status_code=400)
    except Exception as e:
        raise AppException(
            message=f"Failed to get workflow status: {str(e)}",
            status_code=500,
        )

File: frontend/src/api/github.js
import axios from 'axios'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'
const TOKEN_KEY = 'gh_token'

/**
 * Create a pull request with the generated YAML.
 * Uses the GitHub token from localStorage.
 */
export async function createPR(repoUrl, yaml, branchName = 'ci/add-pipeline') {
  try {
    const token = localStorage.getItem(TOKEN_KEY)
    const response = await axios.post(
      `${API_BASE_URL}/api/v1/github/pr`,
      { repo_url: repoUrl, yaml, branch_name: branchName },
      { headers: { Authorization: `Bearer ${token}` } }
    )
    return response.data
  } catch (err) {
    const message = err.response?.data?.message || 'Failed to create pull request. Please try again.'
    throw new Error(message)
  }
}

/**
 * Poll the status of a GitHub Actions workflow run.
 * Requires both the run ID and the repo URL.
 */
export async function pollStatus(runId, repoUrl) {
  try {
    const token = localStorage.getItem(TOKEN_KEY)
    const response = await axios.get(
      `${API_BASE_URL}/api/v1/github/status/${runId}`,
      {
        params: { repo_url: repoUrl },
        headers: { Authorization: `Bearer ${token}` },
      }
    )
    return response.data
  } catch (err) {
    const status = err.response?.status
    const message = err.response?.data?.message

    if (status === 404) {
      throw new Error(message || 'Workflow run not found. The run ID may be invalid.')
    }

    throw new Error(message || 'Failed to fetch workflow status.')
  }
}
